plugin manifest issues

As explained in another issue/post, I create my plugins not as ZIP but as a single jar containing all dependencies.

For that I use the following build-plugin with maven:

```
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-assembly-plugin</artifactId>
                <version>2.5.5</version>

                <executions>
                    <execution>
                        <id>make-plugin-assembly</id>
                        <!-- extend phase package to assembly the archives -->
                        <phase>package</phase>
                        <goals>
                            <goal>single</goal>
                        </goals>
                        <configuration>
                            <archive>
                                <manifestEntries>
                                    <Plugin-Id>${plugin.id}</Plugin-Id>
                                    <Plugin-Class>${plugin.class}</Plugin-Class>
                                    <Plugin-Version>${plugin.version}</Plugin-Version>
                                    <Plugin-Provider>${plugin.provider}</Plugin-Provider>
                                    <Plugin-Dependencies>${plugin.dependencies}</Plugin-Dependencies>
                                </manifestEntries>
                            </archive>
                            <descriptorRefs>
                                <descriptorRef>jar-with-dependencies</descriptorRef>
                            </descriptorRefs>
                            <finalName>${project.build.finalName}.plugin</finalName>
                            <appendAssemblyId>false</appendAssemblyId>
                        </configuration>
                    </execution>
                </executions>
            </plugin>	
```

Works quite well. Now I wanted to have the plugin-properties not in pom.xml

```
    <properties>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
        <maven.compiler.source>1.7</maven.compiler.source>
        <maven.compiler.target>1.7</maven.compiler.target>

        <plugin.id>de.root1.kad-logicplugin</plugin.id>
        <plugin.class>de.root1.kad.logicplugin.LogicPlugin</plugin.class>
        <plugin.version>1.0.0</plugin.version>
        <plugin.provider>Alexander Christian</plugin.provider>
        <plugin.dependencies></plugin.dependencies>        
    </properties>
```

but in an external plugin.properties file (makes life easier when using development-runtimemode). To not need to define plugin-properties twice (pom + properties), I used the following maven-build-plugin to read it from properties-file and use it in pom:


```
            <plugin>
                <groupId>org.codehaus.mojo</groupId>
                <artifactId>properties-maven-plugin</artifactId>
                <version>1.0-alpha-2</version>
                <executions>
                    <execution>
                        <id>read-plugin-properties</id>
                        <phase>initialize</phase>
                        <goals>
                            <goal>read-project-properties</goal>
                        </goals>
                    </execution>
                </executions>
                <configuration>
                    <files>
                        <file>${basedir}/plugin.properties</file>
                    </files>
                </configuration>
            </plugin>
```

Works in principle, but creates an name-conflict... It tooks me 2hrs to find the root-cause:

The pom variable ${plugin.id} is not only used by pf4j/it's plugin-configuration, but also by the maven assembly plugin... BUT: This only get's visible when reading the properties from properties-file. As long as you define the properties directly in pom.xml, ${plugin.id} will have the value we defined by outselves, don't matter where in the pom-file you use it. 
If the property is read from propert-file (not direclty defined in pom), ${plugin.id} will contain the plugin-id of the build-plugin you're currenty configuring. In this case, it is "org.apache.maven.plugins:maven-assembly-plugin:2.5.5".

And it's getting further:

With ${plugin.class} you get an exception when running maven build:

```[ERROR] Failed to execute goal org.apache.maven.plugins:maven-assembly-plugin:2.5.5:single (make-plugin-assembly) on project logicplugin: A type incompatibility occured while executing org.apache.maven.plugins:maven-assembly-plugin:2.5.5:single: java.lang.Class cannot be cast to java.lang.String```

So ${plugin.class} at this point is not containing the plugin-class-string, but the plugin class object...

Maybe this is a kind of special case, as most users will not do that "advanced stuff" that I do.
But I recommend anyway to not use conflicting names.
Maybe prefix all the plugin-property-names with "pf4j." ....

I tried to overcome this issue with having my own manifestPluginDescriptorFinder... But again, some "private" tags prevent me from doing this... :-(

The ManifestPluginDescriptorFinder itself can be overwritten, but it depends on the Plugindescriptor, which has the "setXYZ()" methods package-private :-(

I would overcome also this by overwriting it and having a large constructor instead of using all the package-private setter... But this would be more a hack than "good code". 

So, finally, two things should change:

1) prefix the plugin-properties with "pf4j."
2) enable the user to have an own prefix by overwriting *PluginDescriptorFinder (which is in principle already possible) and enable access to PluginDescriptor to set (somehow) the required values (so that overwriting *PlugindescriptorFinder makes sense.

br,
Alex

--- a/pf4j/src/main/java/ro/fortsoft/pf4j/ManifestPluginDescriptorFinder.java
+++ b/pf4j/src/main/java/ro/fortsoft/pf4j/ManifestPluginDescriptorFinder.java
@@ -13,17 +13,16 @@
 package ro.fortsoft.pf4j;
 
 import com.github.zafarkhaja.semver.Version;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import ro.fortsoft.pf4j.util.StringUtils;
+
 import java.io.File;
 import java.io.FileInputStream;
 import java.io.FileNotFoundException;
 import java.io.IOException;
 import java.util.jar.Attributes;
 import java.util.jar.Manifest;
-
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
-import ro.fortsoft.pf4j.util.StringUtils;
 
 /**
  * Read the plugin descriptor from the manifest file.
@@ -42,74 +41,87 @@ public class ManifestPluginDescriptorFinder implements PluginDescriptorFinder {
 
 	@Override
 	public PluginDescriptor find(File pluginRepository) throws PluginException {
-    	// TODO it's ok with first classes directory? Another idea is to specify in PluginClasspath the folder.
-		String classes = pluginClasspath.getClassesDirectories().get(0);
+        Manifest manifest = readManifest(pluginRepository);
+
+        PluginDescriptor pluginDescriptor = createPluginDescriptor(manifest);
+        validatePluginDescriptor(pluginDescriptor);
+
+		return pluginDescriptor;
+	}
+
+    protected Manifest readManifest(File pluginRepository) throws PluginException {
+        // TODO it's ok with first classes directory? Another idea is to specify in PluginClasspath the folder.
+        String classes = pluginClasspath.getClassesDirectories().get(0);
         File manifestFile = new File(pluginRepository, classes + "/META-INF/MANIFEST.MF");
         log.debug("Lookup plugin descriptor in '{}'", manifestFile);
         if (!manifestFile.exists()) {
             throw new PluginException("Cannot find '" + manifestFile + "' file");
         }
 
-    	FileInputStream input = null;
-		try {
-			input = new FileInputStream(manifestFile);
-		} catch (FileNotFoundException e) {
-			// not happening
-		}
+        FileInputStream input = null;
+        try {
+            input = new FileInputStream(manifestFile);
+        } catch (FileNotFoundException e) {
+            // not happening
+        }
 
-    	Manifest manifest = null;
+        Manifest manifest = null;
         try {
             manifest = new Manifest(input);
         } catch (IOException e) {
             throw new PluginException(e.getMessage(), e);
         } finally {
             try {
-				input.close();
-			} catch (IOException e) {
-				throw new PluginException(e.getMessage(), e);
-			}
+                input.close();
+            } catch (IOException e) {
+                throw new PluginException(e.getMessage(), e);
+            }
         }
 
+        return manifest;
+    }
+
+    protected PluginDescriptor createPluginDescriptor(Manifest manifest) {
         PluginDescriptor pluginDescriptor = new PluginDescriptor();
 
         // TODO validate !!!
-        Attributes attrs = manifest.getMainAttributes();
-        String id = attrs.getValue("Plugin-Id");
-        if (StringUtils.isEmpty(id)) {
-        	throw new PluginException("Plugin-Id cannot be empty");
-        }
+        Attributes attributes = manifest.getMainAttributes();
+        String id = attributes.getValue("Plugin-Id");
         pluginDescriptor.setPluginId(id);
 
-        String description = attrs.getValue("Plugin-Description");
+        String description = attributes.getValue("Plugin-Description");
         if (StringUtils.isEmpty(description)) {
-        	pluginDescriptor.setPluginDescription("");
+            pluginDescriptor.setPluginDescription("");
         } else {
             pluginDescriptor.setPluginDescription(description);
         }
 
-        String clazz = attrs.getValue("Plugin-Class");
-        if (StringUtils.isEmpty(clazz)) {
-        	throw new PluginException("Plugin-Class cannot be empty");
-        }
+        String clazz = attributes.getValue("Plugin-Class");
         pluginDescriptor.setPluginClass(clazz);
 
-        String version = attrs.getValue("Plugin-Version");
-        if (StringUtils.isEmpty(version)) {
-        	throw new PluginException("Plugin-Version cannot be empty");
-        }
+        String version = attributes.getValue("Plugin-Version");
         pluginDescriptor.setPluginVersion(Version.valueOf(version));
 
-        String provider = attrs.getValue("Plugin-Provider");
+        String provider = attributes.getValue("Plugin-Provider");
         pluginDescriptor.setProvider(provider);
-        String dependencies = attrs.getValue("Plugin-Dependencies");
+        String dependencies = attributes.getValue("Plugin-Dependencies");
         pluginDescriptor.setDependencies(dependencies);
 
-        String requires = attrs.getValue("Plugin-Requires");
+        String requires = attributes.getValue("Plugin-Requires");
         if (StringUtils.isNotEmpty(requires)) {
-        	pluginDescriptor.setRequires(requires);
+            pluginDescriptor.setRequires(requires);
         }
 
-		return pluginDescriptor;
-	}
+        return pluginDescriptor;
+    }
+
+    protected void validatePluginDescriptor(PluginDescriptor pluginDescriptor) throws PluginException {
+        if (StringUtils.isEmpty(pluginDescriptor.getPluginId())) {
+            throw new PluginException("Plugin-Id cannot be empty");
+        }
+        if (StringUtils.isEmpty(pluginDescriptor.getPluginClass())) {
+            throw new PluginException("Plugin-Class cannot be empty");
+        }
+    }
 
 }

--- a/pf4j/src/main/java/ro/fortsoft/pf4j/ManifestPluginDescriptorFinder.java
+++ b/pf4j/src/main/java/ro/fortsoft/pf4j/ManifestPluginDescriptorFinder.java
@@ -100,7 +100,9 @@ public class ManifestPluginDescriptorFinder implements PluginDescriptorFinder {
         pluginDescriptor.setPluginClass(clazz);
 
         String version = attributes.getValue("Plugin-Version");
-        pluginDescriptor.setPluginVersion(Version.valueOf(version));
+        if (StringUtils.isNotEmpty(version)) {
+            pluginDescriptor.setPluginVersion(Version.valueOf(version));
+        }
 
         String provider = attributes.getValue("Plugin-Provider");
         pluginDescriptor.setProvider(provider);
@@ -121,6 +123,9 @@ public class ManifestPluginDescriptorFinder implements PluginDescriptorFinder {
         }
         if (StringUtils.isEmpty(pluginDescriptor.getPluginClass())) {
             throw new PluginException("Plugin-Class cannot be empty");
+        }
+        if (pluginDescriptor.getVersion() == null) {
+            throw new PluginException("Plugin-Version cannot be empty");
         }
     }
 

--- a/pf4j/src/main/java/ro/fortsoft/pf4j/PropertiesPluginDescriptorFinder.java
+++ b/pf4j/src/main/java/ro/fortsoft/pf4j/PropertiesPluginDescriptorFinder.java
@@ -96,7 +96,9 @@ public class PropertiesPluginDescriptorFinder implements PluginDescriptorFinder 
         pluginDescriptor.setPluginClass(clazz);
 
         String version = properties.getProperty("plugin.version");
-        pluginDescriptor.setPluginVersion(Version.valueOf(version));
+        if (StringUtils.isNotEmpty(version)) {
+            pluginDescriptor.setPluginVersion(Version.valueOf(version));
+        }
 
         String provider = properties.getProperty("plugin.provider");
         pluginDescriptor.setProvider(provider);
@@ -110,10 +112,12 @@ public class PropertiesPluginDescriptorFinder implements PluginDescriptorFinder 
     protected void validatePluginDescriptor(PluginDescriptor pluginDescriptor) throws PluginException {
         if (StringUtils.isEmpty(pluginDescriptor.getPluginId())) {
             throw new PluginException("plugin.id cannot be empty");
-
         }
         if (StringUtils.isEmpty(pluginDescriptor.getPluginClass())) {
             throw new PluginException("plugin.class cannot be empty");
+        }
+        if (pluginDescriptor.getVersion() == null) {
+            throw new PluginException("plugin.version cannot be empty");
         }
     }
 

--- a/pf4j/src/main/java/ro/fortsoft/pf4j/PropertiesPluginDescriptorFinder.java
+++ b/pf4j/src/main/java/ro/fortsoft/pf4j/PropertiesPluginDescriptorFinder.java
@@ -13,17 +13,16 @@
 package ro.fortsoft.pf4j;
 
 import com.github.zafarkhaja.semver.Version;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import ro.fortsoft.pf4j.util.StringUtils;
+
 import java.io.File;
 import java.io.FileInputStream;
 import java.io.FileNotFoundException;
 import java.io.IOException;
 import java.io.InputStream;
 import java.util.Properties;
-
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
-import ro.fortsoft.pf4j.util.StringUtils;
 
 /**
  * Find a plugin descriptor in a properties file (in plugin repository).
@@ -48,59 +47,74 @@ public class PropertiesPluginDescriptorFinder implements PluginDescriptorFinder 
 
 	@Override
 	public PluginDescriptor find(File pluginRepository) throws PluginException {
+        Properties properties = readProperties(pluginRepository);
+
+        PluginDescriptor pluginDescriptor = createPluginDescriptor(properties);
+        validatePluginDescriptor(pluginDescriptor);
+
+        return pluginDescriptor;
+	}
+
+    protected Properties readProperties(File pluginRepository) throws PluginException {
         File propertiesFile = new File(pluginRepository, propertiesFileName);
         log.debug("Lookup plugin descriptor in '{}'", propertiesFile);
         if (!propertiesFile.exists()) {
             throw new PluginException("Cannot find '" + propertiesFile + "' file");
         }
 
-    	InputStream input = null;
-		try {
-			input = new FileInputStream(propertiesFile);
-		} catch (FileNotFoundException e) {
-			// not happening
-		}
-
-    	Properties properties = new Properties();
+        InputStream input = null;
         try {
-        	properties.load(input);
+            input = new FileInputStream(propertiesFile);
+        } catch (FileNotFoundException e) {
+            // not happening
+        }
+
+        Properties properties = new Properties();
+        try {
+            properties.load(input);
         } catch (IOException e) {
             throw new PluginException(e.getMessage(), e);
         } finally {
             try {
-				input.close();
-			} catch (IOException e) {
-				throw new PluginException(e.getMessage(), e);
-			}
+                input.close();
+            } catch (IOException e) {
+                throw new PluginException(e.getMessage(), e);
+            }
         }
 
+        return properties;
+    }
+
+    protected PluginDescriptor createPluginDescriptor(Properties properties) {
         PluginDescriptor pluginDescriptor = new PluginDescriptor();
 
         // TODO validate !!!
         String id = properties.getProperty("plugin.id");
-        if (StringUtils.isEmpty(id)) {
-        	throw new PluginException("plugin.id cannot be empty");
-        }
         pluginDescriptor.setPluginId(id);
 
         String clazz = properties.getProperty("plugin.class");
-        if (StringUtils.isEmpty(clazz)) {
-        	throw new PluginException("plugin.class cannot be empty");
-        }
         pluginDescriptor.setPluginClass(clazz);
 
         String version = properties.getProperty("plugin.version");
-        if (StringUtils.isEmpty(version)) {
-        	throw new PluginException("plugin.version cannot be empty");
-        }
         pluginDescriptor.setPluginVersion(Version.valueOf(version));
 
         String provider = properties.getProperty("plugin.provider");
         pluginDescriptor.setProvider(provider);
+
         String dependencies = properties.getProperty("plugin.dependencies");
         pluginDescriptor.setDependencies(dependencies);
 
-		return pluginDescriptor;
-	}
+        return pluginDescriptor;
+    }
+
+    protected void validatePluginDescriptor(PluginDescriptor pluginDescriptor) throws PluginException {
+        if (StringUtils.isEmpty(pluginDescriptor.getPluginId())) {
+            throw new PluginException("plugin.id cannot be empty");
+
+        }
+        if (StringUtils.isEmpty(pluginDescriptor.getPluginClass())) {
+            throw new PluginException("plugin.class cannot be empty");
+        }
+    }
 
 }

